proposals/self-types.md

```diff
@@ -0,0 +1,460 @@
+# Self types
+
+* **Type**: Design proposal
+* **Authors**: Maksim Grankin
+* **Status**: Prototype implemented
+* **Issue**: [KT-6494](https://youtrack.jetbrains.com/issue/KT-6494)
+
+
+A **Self type** is type that refers to the receiver type.
+
+```kotlin
+open class A {
+    fun foo(): Self {
+        return this;
+    }
+}
+
+class B : A {
+    
```

Comment:
```kotlin
class B: A {
  fun bar():Self {
     return this;
  }
}
```

Reply:
We do not need `bar` to return `this`.

`foo` from class `A` returns `Self` type.
When we call `foo` for type `B` it will return type `B`.

This is basically the point of having self-types. Take a look at the transformation chains example.

Reply:
i mean we can call bar with self type, even the bar is defind on b, I think that's would be a better example.

Reply:
I see, thats right!

I will update example

Reply:
Done, thanks!